[JSC] Implement Array#concat in C++

Constellation opened the pull request:
#### 184275291379e0cb2d6a9f87e11c0c3a623418a3
<pre>
[JSC] Implement Array#concat in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=280957">https://bugs.webkit.org/show_bug.cgi?id=280957</a>
<a href="https://rdar.apple.com/137404542">rdar://137404542</a>

Reviewed by Yijia Huang.

Let&apos;s move isConcatSpreadable check from tryGetById based on in JS to
watchpoint basis. As a result, we can efficiently implement entire
Array#concat in C++. This is great since we can keep it really fast even
for Baseline level execution (FTL JS is so fast, but Baseline is not.
And we would like to make builtin functions fast even if it is not
executed enough yet).

We remove JS concat and concatSlowPath, and C++ concatMemcpy. And move
everything to C++ Array#concat.

                                                         ToT                     Patched

    array-prototype-concat-copy-double-and-int32
                                                    1.3390+-0.0116     ^      0.8628+-0.0153        ^ definitely 1.5519x faster
    array-prototype-concat-copy-int32               1.0441+-0.0092     ^      0.6616+-0.0088        ^ definitely 1.5780x faster
    array-prototype-concat-copy-obj                 1.0724+-0.0132     ^      0.6849+-0.0085        ^ definitely 1.5657x faster
    array-concat-int32-and-double                   0.9596+-0.0107     ^      0.9139+-0.0128        ^ definitely 1.0500x faster
    array-concat-double-and-int32                   0.9622+-0.0111     ^      0.9141+-0.0142        ^ definitely 1.0527x faster
    array-prototype-concat-copy-double              1.0813+-0.0111     ^      0.6855+-0.0102        ^ definitely 1.5774x faster
    concat-append-one                               2.5387+-0.0189     ^      2.4526+-0.0242        ^ definitely 1.0351x faster

* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(linkTimeConstant.concatSlowPath): Deleted.
(concat): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::concatAppendOne):
(JSC::concatAppendArray):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/ArrayPrototypeInlines.h:
(JSC::arrayMissingIsConcatSpreadable):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::arrayIsConcatSpreadableWatchpointSet):

Canonical link: <a href="https://commits.webkit.org/284847@main">https://commits.webkit.org/284847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9284eedce7be6d93dff674ad6979c32ada200442

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70595 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74684 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21773 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57801 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21613 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55917 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14385 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45466 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60857 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36379 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42123 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18293 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20134 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63716 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76404 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69843 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17860 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63641 "Found 4 new test failures: compositing/animation/keyframe-order.html compositing/repaint/copy-forward-dirty-region-purged-backbuffer.html fast/text/crash-letter-spacing-infinite.html imported/w3c/web-platform-tests/css/css-view-transitions/inline-element-size.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14864 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63579 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11628 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5274 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91625 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10839 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45803 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/571 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/19975 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46877 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48154 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46619 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 reviewed (approved):
Nice, r=me if bots are happy.

Constellation committed:
[JSC] Implement Array#concat in C++
https://bugs.webkit.org/show_bug.cgi?id=280957
rdar://137404542

Reviewed by Yijia Huang.

Let's move isConcatSpreadable check from tryGetById based on in JS to
watchpoint basis. As a result, we can efficiently implement entire
Array#concat in C++. This is great since we can keep it really fast even
for Baseline level execution (FTL JS is so fast, but Baseline is not.
And we would like to make builtin functions fast even if it is not
executed enough yet).

We remove JS concat and concatSlowPath, and C++ concatMemcpy. And move
everything to C++ Array#concat.

                                                         ToT                     Patched

    array-prototype-concat-copy-double-and-int32
                                                    1.3390+-0.0116     ^      0.8628+-0.0153        ^ definitely 1.5519x faster
    array-prototype-concat-copy-int32               1.0441+-0.0092     ^      0.6616+-0.0088        ^ definitely 1.5780x faster
    array-prototype-concat-copy-obj                 1.0724+-0.0132     ^      0.6849+-0.0085        ^ definitely 1.5657x faster
    array-concat-int32-and-double                   0.9596+-0.0107     ^      0.9139+-0.0128        ^ definitely 1.0500x faster
    array-concat-double-and-int32                   0.9622+-0.0111     ^      0.9141+-0.0142        ^ definitely 1.0527x faster
    array-prototype-concat-copy-double              1.0813+-0.0111     ^      0.6855+-0.0102        ^ definitely 1.5774x faster
    concat-append-one                               2.5387+-0.0189     ^      2.4526+-0.0242        ^ definitely 1.0351x faster

* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(linkTimeConstant.concatSlowPath): Deleted.
(concat): Deleted.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::concatAppendOne):
(JSC::concatAppendArray):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ArrayPrototype.h:
* Source/JavaScriptCore/runtime/ArrayPrototypeInlines.h:
(JSC::arrayMissingIsConcatSpreadable):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::arrayIsConcatSpreadableWatchpointSet):

Canonical link: https://commits.webkit.org/284847@main
